lang/machine.md

```diff
@@ -20,6 +20,18 @@ pub struct Machine<M: Memory> {
 
     /// The stack.
     stack: List<StackFrame<M>>,
+
+    /// Stores a pointer to each of the global allocations.
+    global_ptrs: Map<GlobalName, Pointer<M::Provenance>>,
+
+    /// Stores an address for each function name.
+    fn_addrs: Map<FnName, mem::Address>,
+
+    /// Maps address of functions to the actual function
+    fn_map: Map<mem::Address, Function>,
```

Comment:
Is there a method `Map.search` or should I just use a for loop to iterate over all of them?

Reply:
The maps aren't really inverse of each other since one works with `FnName` and the other with `Function`. I think if we look at redundancy then there is redundancy in the set `prog`, `fn_addrs`, and `fn_map`. 

Of course if the Program is needed for anything different than translating `FnName` to `Function` removing `fn_map` is the right choice. 
I prefer to remove `prog` from the Machine, but since I probably overlooked something I commited the removal of `fn_map`.

Reply:
> I prefer to remove prog from the Machine

I guess the only part of `prog` that we still need is `prog.functions`, so that would be a possibility, yeah -- good point. Let's not do this in the same PR though.